Add Command Palette integration for ActivityPub

obenland opened the pull request:
[The Command Palette will become available in all of wp-admin in WordPress 6.9](https://make.wordpress.org/core/2025/07/28/roadmap-to-6-9/#command-palette-everywhere), this PR adds a few common navigation commands:

<img width="1510" height="985" alt="Screenshot 2025-10-14 at 12 01 13 PM" src="https://github.com/user-attachments/assets/3a5d9040-c40e-488a-a63d-bfb8b817d1b0" />

## Proposed changes:
* Add integration with WordPress Command Palette (Cmd/Ctrl + K) for quick navigation to ActivityPub admin pages
* Register commands that adapt based on actor mode (actor/blog/actor_blog) and user capabilities
* Add TypeScript support with proper type declarations for @wordpress/commands package
* Include ActivityPub icon for command palette entries
* Add dynamic command loader for searching and editing extra fields

### Commands added:
- **View User/Blog Followers** - Navigate to followers list
- **View User/Blog Following** - Navigate to following list (when enabled)
- **View Blocked Actors** - Navigate to blocked actors list
- **View ActivityPub Settings** - Navigate to settings page (admin only)
- **View Extra Fields** - Navigate to extra fields list
- **Add New Extra Field** - Navigate to create new extra field
- **Edit Extra Field** - Dynamic search to quickly edit any of your extra fields

### Other information:

- [x] Have you written new tests for your changes, if applicable? - N/A (UI integration)

## Testing instructions:

1. Install and activate the ActivityPub plugin with this branch.
2. Go to the Site Editor.
3. Press **Cmd+K** (Mac) or **Ctrl+K** (Windows/Linux) to open the Command Palette.
4. Type "ActivityPub" to see all available commands.
5. Verify commands appear based on your actor mode and capabilities:
   - User Followers/Following commands appear when actor mode includes 'actor'
   - Blog Followers/Following commands appear when actor mode includes 'blog' and user has manage_options capability
   - Settings command only appears for users with manage_options capability
6. Click any command to verify it navigates to the correct admin page.
7. Type the name of an extra field to see dynamic "Edit" commands appear for your extra fields.
8. Click an edit command to navigate directly to editing that extra field.

### Changelog entry

- [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

- [x] Minor

#### Type

- [x] Added - for new features

#### Message

Add Command Palette integration for quick navigation to ActivityPub admin pages

</details>

obenland committed:
Add Command Palette integration for ActivityPub

- Register ActivityPub commands in WordPress Command Palette (Cmd/Ctrl + K)
- Add commands for navigating to Followers, Following, Blocked Actors, Settings, and Extra Fields
- Commands adapt based on actor mode (actor/blog/actor_blog) and user capabilities
- Add TypeScript support with proper type declarations for @wordpress/commands
- Include ActivityPub icon for command palette entries
- Add @wordpress/commands as a dependency

Commands include:
- View User/Blog Followers
- View User/Blog Following (when enabled)
- View Blocked Actors
- View ActivityPub Settings (admin only)
- View Extra Fields
- Add New Extra Field

matticbot committed:
Add changelog

obenland committed:
Use document.location instead of document.location.href

Match WordPress core convention for navigation

obenland commented:
I wanted to add the ability to edit individual Extra Fields directly but ran into a roadblock with the lack of author information for them.

obenland committed:
Add dynamic command loader for editing extra fields

- Add useCommandLoader hook to enable searching and editing extra fields
- Filter extra fields by current user author
- Use sprintf for proper i18n with placeholder
- Add author support to extra fields post type
- Restrict editing others' user extra fields with capabilities
- Remove pre_get_posts filter (now handled by capabilities)

obenland commented:
@pfefferle I'm curious what you think about the changes in [the extrafields post type registration](https://github.com/Automattic/wordpress-activitypub/pull/2315/files#diff-ceb59dc4107c120554c144aa5f601712077313b8b65bc0a0a844b0125272461f) and [the pre_get_posts filter](https://github.com/Automattic/wordpress-activitypub/pull/2315/files#diff-55e60d9ff58d38ea1c3b83c950a6fc45a8b75fc8bb1ce6198409bf0dc7132f31L409-L418).

obenland committed:
Add unit tests for command palette plugin

- Test configuration object and actor modes
- Test sprintf utility for i18n
- Test title sanitization (quote removal)
- Test command name generation with unique IDs
- Test navigation URL generation
- 18 tests covering core functionality

obenland committed:
Add @types/jest for TypeScript support in tests

- Install @types/jest package
- Add jest to types field in tsconfig.json
- Resolves TypeScript errors in test files

pfefferle commented:
When I change the Actor mode to Blog-Only, I can't see any ActivityPub Actions, only in "Actors" or "Actors and Blog" modes.

pfefferle commented:
@obenland are you sure that the command palette is coming in WP 6.9? I tested the latest nightly build and could not use it outside the site-editor!?

obenland commented:
It's in Gutenberg, it might not have been merged to Core yet.

obenland committed:
Fix command palette availability for all actor modes

- Reorganize commands by actor mode:
  - Actor mode: User-specific commands (followers, extra fields, blocked actors)
  - Blog mode: Blog-specific commands (blog followers, settings)
- Enqueue script for users with manage_options OR activitypub capability
- Ensures blog-only mode users with manage_options see commands
- Add comprehensive documentation of command availability logic

obenland committed:
Allow admins to access command palette integration

Expanded the condition to enqueue the command palette integration for users with 'manage_options' capability, enabling administrators to access this feature in addition to users with ActivityPub permissions.

obenland committed:
Add blog extra fields commands for blog mode

- Add 'View Blog Extra Fields' command
- Add 'Add New Blog Extra Field' command
- Add dynamic command loader for editing blog extra fields
- Blog extra fields use ap_extrafield_blog post type
- Commands only available in blog/actor_blog mode with manage_options
- Update documentation to include new commands

obenland commented:
> When I change the Actor mode to Blog-Only, I can't see any ActivityPub Actions, only in "Actors" or "Actors and Blog" modes.

@pfefferle Thanks for that feedback, this should now be fixed and ready for another round of reviews

pfefferle commented:
When I install Gutenberg, I can call the command palette on every site, but the ActivityPub commands only show up in the site-editor.

That is quite confusing, because I would expect it the other way around.

Maybe we should wait before the feature is released?

obenland commented:
I wonder if it's a matter of dependencies not being enqueued properly, I'll have another look later today.

obenland committed:
Refactor command palette to work in all wp-admin pages

Previously, commands were registered using `registerPlugin` which only
works in editor contexts (post editor, site editor). This refactor
switches to direct command registration using `dispatch('core/commands')`
which works throughout wp-admin.

Key changes:
- Replace `registerPlugin` + `useCommand` with direct `dispatch` calls
- Use `registerCommand()` and `registerCommandLoader()` helper functions
- Move all type definitions to `types.d.ts` for better organization
- Add proper TypeScript types using WordPress core types
- Use `Post` type from `@wordpress/core-data` for entity records
- Minimize `any` types (only for `hasFinishedResolution` metadata)

Commands now appear in the command palette (Cmd+K) across all wp-admin
pages, not just in the site editor.

obenland commented:
> When I install Gutenberg, I can call the command palette on every site, but the ActivityPub commands only show up in the site-editor.

Nice catch! Previously, commands were registered using `registerPlugin` which only works in editor contexts (post editor, site editor). [c0d87da](https://github.com/Automattic/wordpress-activitypub/pull/2315/commits/c0d87dae6fd1f0de564a092299eb7635dd1abaa0) switches to direct command registration which works throughout wp-admin.

This should be ready for another review.

pfefferle commented on src/command-palette/plugin.tsx:
```diff
@@ -0,0 +1,303 @@
+	if ( followingEnabled ) {
+		registerCommand( {
+			name: 'activitypub/navigate-blog-following',
+			label: __( 'ActivityPub: View Blog Following', 'activitypub' ),
```
Let's use a similar wording for "View Who You Follow" and "View Blog Following", to make selections easier. Either use "View your Followings" or "View Who your Blog Follows" (<- or something that makes more sense in english 😉)

<img width="498" height="209" alt="Image" src="https://github.com/user-attachments/assets/5679d059-0022-460d-9ea0-9df8aefd1035" />

obenland replied:
Fixed in [c755f21](https://github.com/Automattic/wordpress-activitypub/pull/2315/commits/c755f21520bc71ae5595fcd343f1862aad095cf3).


<img width="418" height="401" alt="Screenshot 2025-10-16 at 1 21 57 PM" src="https://github.com/user-attachments/assets/ae932e77-8e93-4164-ad05-fdaa904536b0" />

pfefferle commented on src/command-palette/types.d.ts:
```diff
@@ -0,0 +1,97 @@
+ */
+export interface ActivityPubCommandPaletteConfig {
+	followingEnabled: boolean;
+	actorMode: 'actor' | 'blog' | 'actor_blog';
```
Just a note: We should find a way, to share consts between PHP and JS. The direct use of the values is a bit fragile.

pfefferle commented:
What do you think about introducing a simpler, more general way to distinguish between User and Blog settings?

For example:
`ActivityPub: Blog > View Followers`
`ActivityPub: User > View Followers`

This would make filtering by actor types easier, you could just type `ActivityPub: Blog` to see all Blog related commands or `View Followers` to see all Followers related commands.
I’m not entirely sure about the separator, though, since `>` is usually used for paths or breadcrumbs. We could consider using `-` or `~` instead.

obenland committed:
Update blog following label

obenland committed:
Remove unused dependency

obenland commented:
> What do you think about introducing a simpler, more general way to distinguish between User and Blog settings?

I'm not sure it's necessary? I worry it's a separation we think more about on a technical level than how it's perceived by others. "View Who You Follow" feels much more personable than "User > View Followers".

pfefferle commented:
My point was the unification of the wording, not so much the separator thing, I simply used a shorter version. My point is that I do not like the different wording for user and blog versions.

obenland commented:
Gotcha, that should be fixed with c755f21520bc71ae5595fcd343f1862aad095cf3.

pfefferle committed:
Merge branch 'trunk' into add/command-palette

obenland committed:
Revert c755f21520bc71ae5595fcd343f1862aad095cf3

See https://github.com/Automattic/wordpress-activitypub/commit/c755f21520bc71ae5595fcd343f1862aad095cf3#r168132552
